### PR TITLE
Fix SetBucketPolicy Functional test

### PR DESCRIPTION
**Pull request by kannappanr**
Set the time delay to 15 seconds
Set header 'x-amz-acl' to 'bucket-owner-full-control', while creating
the object anonymously, so it can be downloaded by bucket owner on AWS S3.
source: http://prettyplease.me/anonymous-s3-upload-with-full-owner-control/

### CODE REVIEW

File: functional/FunctionalTest.java [hunk 1]
```diff
@@ -2421,16 +2421,9 @@ public static void testSetBucketPolicy(String objectPrefix, PolicyType policyTyp
         client.putObject(bucketName, objectName, new ByteArrayInputStream(data), data.length, customContentType);
       }
 
+      // Authenticated download is not supported in AWS S3.
```
Comment (balamurugana) [inline functional/FunctionalTest.java]: Unable to understand this comment.

File: functional/FunctionalTest.java [hunk 2]
```diff
@@ -2421,16 +2421,9 @@ public static void testSetBucketPolicy(String objectPrefix, PolicyType policyTyp
...
       if ((policyType == PolicyType.READ_WRITE) || (policyType == PolicyType.WRITE_ONLY)) {
         writeObject(urlString, data);
-
-        InputStream is = client.getObject(bucketName, objectName);
```
Comment (balamurugana) [inline functional/FunctionalTest.java]: This check is doing whether anonymously uploaded object by `writeObject()` is downloadable by the authenticated user and its same content or not.

Could you tell me what exactly the problem is?

Comment (kannappanr) [reply functional/FunctionalTest.java]: It is not supported in AWS.

Comment (balamurugana) [reply functional/FunctionalTest.java]: How come an authenticated user not able to download objects whatever the way it is uploaded? I am sure the policy is wrongly set.  Can you confirm?

File: functional/FunctionalTest.java (hunk 1 above)
Comment (balamurugana) [reply functional/FunctionalTest.java]: Where do you see it is not supported?  You could give a doc reference

File: functional/FunctionalTest.java (hunk 2 above)
Comment (harshavardhana) [reply functional/FunctionalTest.java]: This is very simple to test

```
$ mc mb s3/bala-123
Bucket created successfully `s3/bala-123`.
$ mc policy public s3/bala-123
Access permission for `s3/bala-123` is set to `public`

$ curl -X PUT -T /etc/issue https://bala-123.s3.amazonaws.com/public.txt
$ curl https://bala-123.s3.amazonaws.com/public.txt
Ubuntu 17.10 \n \l

$ mc cp s3/bala-123/public.txt /tmp
mc: <ERROR> Failed to copy `https://s3.amazonaws.com/bala-123/public.txt`. Insufficient permissions to access this file `https://s3.amazonaws.com/bala-123/public.txt`

$ mc cp --debug s3/bala-123/public.txt /tmp
...
...
mc: <DEBUG> HEAD /public.txt HTTP/1.1
Host: bala-123.s3.amazonaws.com
User-Agent: Minio (linux; amd64) minio-go/4.0.7 mc/DEVELOPMENT.GOGET
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20180205/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20180205T081532Z

mc: <DEBUG> HTTP/1.1 403 Forbidden
Transfer-Encoding: chunked
Content-Type: application/xml
Date: Mon, 05 Feb 2018 08:15:31 GMT
Server: AmazonS3
X-Amz-Id-2: toz5YBQjx/oJrbXibScrFAN87EJBUTpCcLQRzmZOV+RcjCu/74VAaF0FlxkPZkHxuttxajTbHBA=
X-Amz-Request-Id: 129188C6471E3E70


mc: <DEBUG> Response Time:  190.267361ms

mc: <ERROR> Failed to copy `https://s3.amazonaws.com/bala-123/public.txt`. Insufficient permissions to access this file `https://s3.amazonaws.com/bala-123/public.txt`
 (3) cp-main.go:372 cmd.doCopySession(..) Tags: [https://s3.amazonaws.com/bala-123/public.txt]
 (2) common-methods.go:165 cmd.uploadSourceToTargetURL(..) Tags: [https://s3.amazonaws.com/bala-123/public.txt]
 (1) common-methods.go:93 cmd.getSourceStream(..) Tags: [s3, https://s3.amazonaws.com/bala-123/public.txt]
 (0) client-s3.go:926 cmd.(*s3Client).getObjectStat(..)
 Release-Tag:DEVELOPMENT.GOGET | Commit:DEVELOPMENT. | Host:prtsc | OS:linux | Arch:amd64 | Lang:go1.9.3 | Mem:2.6MB/9.5MB | Heap:2.6MB/5.7MB
```

File: functional/FunctionalTest.java (hunk 1 above)
Comment (harshavardhana) [reply functional/FunctionalTest.java]: It is not documented @balamurugana

**Comment by harshavardhana** [comment]
@kannappanr we can keep the test but use the following technique for uploading anonymous objects 

- https://github.com/aws/aws-cli/issues/2967#issuecomment-363026163
- http://prettyplease.me/anonymous-s3-upload-with-full-owner-control/

File: functional/FunctionalTest.java [hunk 3]
```diff
@@ -211,6 +211,10 @@ public static void writeObject(String urlString, byte[] dataBytes) throws Except
     Request request = requestBuilder
         .url(HttpUrl.parse(urlString))
         .method("PUT", RequestBody.create(null, dataBytes))
+        // Set header 'x-amz-acl' to 'bucket-owner-full-control', so objects created
+        // anonymously, can be downloaded by bucket owner.
+        // Source: http://prettyplease.me/anonymous-s3-upload-with-full-owner-control/
+        .addHeader("x-amz-acl","bucket-owner-full-control")
```
Comment (harshavardhana) [inline functional/FunctionalTest.java]: can you add a space after `.addHeader("x-amz-acl", "bucket-owner-full-control")`

File: functional/FunctionalTest.java [hunk 4]
```diff
@@ -211,6 +211,10 @@ public static void writeObject(String urlString, byte[] dataBytes) throws Except
     Request request = requestBuilder
         .url(HttpUrl.parse(urlString))
         .method("PUT", RequestBody.create(null, dataBytes))
+        // Set header 'x-amz-acl' to 'bucket-owner-full-control', so objects created
+        // anonymously, can be downloaded by bucket owner.
+        // Reference: http://prettyplease.me/anonymous-s3-upload-with-full-owner-control/
```
Comment (balamurugana) [inline functional/FunctionalTest.java]: Move this comment to above requestBuilder. Also mention this is applicable to aws s3 only.  The reference link is not required.

File: functional/FunctionalTest.java [hunk 5]
```diff
@@ -2410,8 +2414,8 @@ public static void testSetBucketPolicy(String objectPrefix, PolicyType policyTyp
     long startTime = System.currentTimeMillis();
     try {
       client.setBucketPolicy(bucketName, objectPrefix, policyType);
-      // Wait for 5 seconds for server to set the policy into effect.
-      Thread.sleep(5000);
+      // Wait for 15 seconds for server to set the policy into effect.
+      Thread.sleep(15000);
```
Comment (balamurugana) [inline functional/FunctionalTest.java]: Have you checked 15 seconds wait is minimum required?

Comment (kannappanr) [reply functional/FunctionalTest.java]: The minimum required seems to be around 7 seconds, so doubled that number.

File: functional/FunctionalTest.java (hunk 4 above)
Comment (kannappanr) [reply functional/FunctionalTest.java]: done

**Pull request merged**